### ISSUE TITLE
Sodium crashes NVIDIA GPU after a while

### ISSUE THREAD

**Issue by Yadokingu** [jellysquid3/Sodium#101]
After playing for a while (10-50 minutes) the NVIDIA GPU crashes and with it the game. Sometimes the game just freezes, sometimes it crashes completely. Afterward no programs that rely on the NVIDIA GPU will run until the computer is rebootet.

[Log from latest crash](https://github.com/jellysquid3/sodium-fabric/files/4939685/latest.log)

At the same time I was logging the GPU with GPU-Z, the time stamps line up: [GPU-Z Sensor Log](https://github.com/jellysquid3/sodium-fabric/files/4939696/GPU-Z.Sensor.Log.txt)

[Windows Event log](https://github.com/jellysquid3/sodium-fabric/files/4939703/sodium.crash.nvidia.windows.event.txt)

- Java Version:  happens with JRE 14 and the included JRE 8
- CPU: Core i5-2450M (Sandybridge)
- GPU: Nvidia Geforce GT 630M

No, this has nothing to do with the integrated Intel GPU, it doesn't run the game and it doesn't crash. See here:
![2020-07-17_17 39 13](https://user-images.githubusercontent.com/11133215/87821899-e01b4f80-c870-11ea-985b-e90ead1d38a3.png)

**Comment by smonnnn** [comment]
This crash is caused by your GPU driver, there is nothing that can be done about this on Sodiums side. It's possible that a clean install of your GPU driver fixes this issue, but other than that there's not much that can be done here.

**Comment by jellysquid3** [comment]
Your GPU driver is crashing HARD and taking down the Graphics Subsystem with it. I'm not sure what could cause such a catastrophic fault. The logs do not provide any useful information as to what the fault is, only that one is happening. From my little knowledge in the German language and some online translators, I was able to piece this out from the error:  

> The following information was recovered from the event:
> Unable to recover from a kernel exception. The application must close.

Usually, Windows should perform a Display Driver Reset (DDR) after Timeout Detection and Recovery (TDR), but it's not doing so here. That might be down to the fact that your computer contains two graphics cards, however I'm not experienced with such hardware.

Can you check for any `hs_err_pid****.log` files in your game directory (i.e. `.minecraft`)? I'm doubtful you'll find anything given that the game fails to properly crash sometimes, but it is worth a shot.

**Comment by jellysquid3** [comment]
Worth noting that after the crash, the GPU-Z sensor logs go blank with default values. It's as if the GPU is just disappearing from the system. It would make sense to me that if your desktop is renderered by the integrated GPU (which is often the case, since it's more power efficient) that it wouldn't bring down the entire system, but only software trying to use the NVIDIA GPU.

**Comment by PaintNinja** [comment]
I wonder if the OpenGL 4.3 renderer is exposing some bad driver bugs... Does it still crash if you set it to the OpenGL 3.0 renderer instead?

**Comment by Yadokingu** [comment]
Thanks for the answers. I completely understand if it can't be fixed, but it would be nice to know for sure if there isn't some fix or workaround.

> Usually, Windows should perform a Display Driver Reset (DDR) after Timeout Detection and Recovery (TDR), but it's not doing so here.

I think it's trying (and thinks it has done it) it just fails. "Der Anzeigetreiber "nvlddmkm" reagiert nicht mehr und wurde wiederhergestellt." means "The display driver "nvlddmkm" isn't reacting anymore and was restored."
[Here is the evtx of the events. It should have english included.](https://github.com/jellysquid3/sodium-fabric/files/4939947/nvidia.crash.zip) Sorry for not thinking about the language.

>Can you check for any hs_err_pid****.log files in your game directory (i.e. .minecraft)? I'm doubtful you'll find anything given that the game fails to properly crash sometimes, but it is worth a shot.

Sadly there aren't any.

>I wonder if the OpenGL 4.3 renderer is exposing some bad driver bugs... Does it still crash if you set it to the OpenGL 3.0 renderer instead?

How would I change it?

**Comment by jellysquid3** [comment]
`Options > Video Settings > Advanced > Chunk Renderer`

Setting it to one of the Oneshot renderers (either 2.0 or 3.0) might work around the issue that's happening here.

**Comment by Yadokingu** [comment]
Chosing a different OpenGL version changed nothing. But, I looked at the windows event log again and found an error code I overlooked before "Error code: 3 (subcode 2)". From what I gather, usually you should get that error displayed, but I guess since windows failed to reset the driver, that didn't happen.
Googleing that code reaveals that it is a common error, usually found with Adobe 3D applications, that crashes the graphics driver. The fix is, get this, in the Nvidia control panel change "power management mode" from "optimal performance" to "maximum performance". Totally intuitive and great that the driver doesn't run correctly on standard settings! /s
Anyway. I can't say for 100% sure, as sometimes it would take a long time to crash, but that seems to have fixed it. 
Sorry to have bothered you with a bug in Nvidias drivers, but maybe it can help the next guy.
Edit: Nope, still crashes, just takes a really long time and stressing the system to do so. I guess it might be useable this way. I'm all out of ideas. Otherwise I'll have to go back to vanilla rendering, which would be sad, since Sodium is amazing.

PS: In all the endless testing I did before finding the solution I also tested the internal Intel HD 3000 GPU and that one ran flawlessly. Of course the frame rate was a lot lower than with the Nvidia GPU, where the Nvidia would have 50-60FPS, the Intel would have 20-30, but for people who only have that one that's still pretty good. The game was absolutely playable.

**Comment by mrgrim** [comment]
I experienced similar symptoms running an older driver version, 441.66. First a hard crash, than a CTD. However, my system was always usable after. This might be due to my lack of an iGPU/using a desktop. I upgraded to 451.67 and played for a few hours without issue after. I'll attach the crash log here in case there's any way the issue can be found and worked around.

[crash1-hs_err_pid18980.log](https://github.com/jellysquid3/sodium-fabric/files/4945387/crash1-hs_err_pid18980.log)
